src/Features/Core/Portable/Diagnostics/Analyzers/NamingStyles/EditorConfigNamingStyleParser.SymbolSpec.cs

```diff
@@ -0,0 +1,235 @@
+
+            symbolSpec = new SymbolSpecification(
+                    Guid.NewGuid(),
+                    symbolSpecName,
+                    symbolKindList: applicableKinds,
+                    accessibilityKindList: applicableAccessibilities,
+                    modifiers: requiredModifiers);
+            return true;
+        }
+
+        private static bool TryGetSymbolSpecNameForNamingRule(
+            string namingRuleName,
+            IReadOnlyDictionary<string, object> conventionsDictionary,
+            out string symbolSpecName)
+        {
+            symbolSpecName = null;
+            object result;
+            if (conventionsDictionary.TryGetValue($"dotnet_naming_rule.{namingRuleName}.symbols", out result))
+            {
+                symbolSpecName = (string)result;
+                return true; ;
+            }
+
+            return false;
+        }
+
+        private static SymbolKindOrTypeKind[] GetSymbolsApplicableKinds(
+            string symbolSpecName,
+            IReadOnlyDictionary<string, object> conventionsDictionary)
+        {
+            object result;
+            if (conventionsDictionary.TryGetValue($"dotnet_naming_symbols.{symbolSpecName}.applicable_kinds", out result))
+            {
+                return ParseSymbolKindList((string)result);
+            }
+
+            return Array.Empty<SymbolKindOrTypeKind>();
+        }
+
+        private static SymbolKindOrTypeKind[] ParseSymbolKindList(string symbolSpecApplicableKinds)
+        {
+            var builder = ArrayBuilder<SymbolKindOrTypeKind>.GetInstance();
+
+            if (symbolSpecApplicableKinds == null)
+            {
+                return Array.Empty<SymbolKindOrTypeKind>();
+            }
+
+            if (symbolSpecApplicableKinds.Trim() == "*")
+            {
+                return new[]
+                {
+                    new SymbolKindOrTypeKind(TypeKind.Class),
+                    new SymbolKindOrTypeKind(TypeKind.Struct),
+                    new SymbolKindOrTypeKind(TypeKind.Interface),
+                    new SymbolKindOrTypeKind(TypeKind.Enum),
+                    new SymbolKindOrTypeKind(SymbolKind.Property),
+                    new SymbolKindOrTypeKind(SymbolKind.Method),
+                    new SymbolKindOrTypeKind(SymbolKind.Field),
+                    new SymbolKindOrTypeKind(SymbolKind.Event),
+                    new SymbolKindOrTypeKind(SymbolKind.Namespace),
+                    new SymbolKindOrTypeKind(TypeKind.Delegate),
+                    new SymbolKindOrTypeKind(SymbolKind.TypeParameter)
+                };
+            }
+
+            foreach (var symbolSpecApplicableKind in symbolSpecApplicableKinds.Split(',').Select(x => x.Trim()))
+            {
+                switch (symbolSpecApplicableKind)
+                {
+                    case "class":
+                        return new[] { new SymbolKindOrTypeKind(TypeKind.Class) };
```

Comment:
1. can we use immutable arrays.
2. can we cache these somewhere so we don't keep allocating what are essentially constants.